pkg/beholder/client.go

```diff
@@ -0,0 +1,289 @@
+	factory := func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error) {
+		return otlploggrpc.New(ctx, options...)
+	}
+	return newOtelClient(cfg, errorHandler, factory)
+}
+
+// Used for testing to override the default exporter
+type otlploggrpcFactory func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error)
+
+func newOtelClient(cfg Config, errorHandler errorHandlerFunc, otlploggrpcNew otlploggrpcFactory) (OtelClient, error) {
+	ctx := context.Background()
+	baseResource, err := newOtelResource(cfg)
+	noop := NewNoopClient()
+	if err != nil {
+		return noop, err
+	}
+	creds := insecure.NewCredentials()
+	if !cfg.InsecureConnection && cfg.CACertFile != "" {
+		creds, err = credentials.NewClientTLSFromFile(cfg.CACertFile, "")
+		if err != nil {
+			return noop, err
+		}
+	}
+	sharedLogExporter, err := otlploggrpcNew(
+		ctx,
+		otlploggrpc.WithTLSCredentials(creds),
+		otlploggrpc.WithEndpoint(cfg.OtelExporterGRPCEndpoint),
+	)
+	if err != nil {
+		return noop, err
+	}
+
+	// Logger
+	loggerProcessor := sdklog.NewBatchProcessor(
+		sharedLogExporter,
+		sdklog.WithExportTimeout(cfg.LogExportTimeout), // Default is 30s
+	)
+	loggerAttributes := []attribute.KeyValue{
+		attribute.String("beholder_data_type", "zap_log_message"),
+	}
+	loggerResource, err := sdkresource.Merge(
+		sdkresource.NewSchemaless(loggerAttributes...),
+		baseResource,
+	)
+	if err != nil {
+		return noop, err
+	}
+	loggerProvider := sdklog.NewLoggerProvider(
+		sdklog.WithResource(loggerResource),
+		sdklog.WithProcessor(loggerProcessor),
+	)
+	logger := loggerProvider.Logger(cfg.PackageName)
+
+	// Tracer
+	tracerProvider, err := newTracerProvider(cfg, baseResource, creds)
+	if err != nil {
+		return noop, err
+	}
+	tracer := tracerProvider.Tracer(cfg.PackageName)
+
+	// Meter
+	meterProvider, err := newMeterProvider(cfg, baseResource, creds)
+	if err != nil {
+		return noop, err
+	}
+	meter := meterProvider.Meter(cfg.PackageName)
+
+	// Message Emitter
+	messageLogProcessor := sdklog.NewBatchProcessor(
+		sharedLogExporter,
+		sdklog.WithExportTimeout(cfg.EmitterExportTimeout), // Default is 30s
+	)
+	messageAttributes := []attribute.KeyValue{
+		attribute.String("beholder_data_type", "custom_message"),
+	}
+	messageLoggerResource, err := sdkresource.Merge(
+		sdkresource.NewSchemaless(messageAttributes...),
+		baseResource,
+	)
+	if err != nil {
+		return noop, err
+	}
+	messageLoggerProvider := sdklog.NewLoggerProvider(
+		sdklog.WithResource(messageLoggerResource),
+		sdklog.WithProcessor(messageLogProcessor),
+	)
+	messageLogger := messageLoggerProvider.Logger(cfg.PackageName)
+	messageEmitter := newMessageEmitter(sharedLogExporter, messageLogger)
+
+	setOtelErrorHandler(errorHandler)
+
+	client := OtelClient{cfg, logger, tracer, meter, messageEmitter, loggerProvider, tracerProvider, meterProvider}
+
+	return client, nil
+}
+
+func (c OtelClient) SetGlobals() {
+	// Logger
```

Comment:
these comments are somewhat helpful (the name of the provider is straightforward thankfully), but I think a reader would benefit more from a top level comment explaining what providers `SetGlobals` is expected to set and a link out to relevant docs

Reply:
@patrickhuie19, added comment [here](https://github.com/smartcontractkit/chainlink-common/pull/694/commits/b80e230cd57a43ccf512f67c615aa5913148270b#diff-74ce7f3d5474f67956a5d9ba5779b478b715235a0d33e8826eb99b146664ae4dR168-R175)